pallets/connectors-gateway/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+		_,
+		Blake2_128Concat,
+		Domain,
+		Blake2_128Concat,
+		DomainAddress,
+		(),
+		ValueQuery,
+	>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Router initialization failed.
+		RouterInitFailed,
+
+		/// The origin of the message to be processed is invalid.
+		InvalidMessageOrigin,
+
+		/// The domain is not supported.
+		DomainNotSupported,
+
+		/// Message decoding error.
+		MessageDecode,
+
+		/// Connector was already added to the domain.
+		ConnectorAlreadyAdded,
+
+		/// Maximum number of connectors for a domain was reached.
+		MaxConnectorsReached,
+
+		/// Connector was not found.
+		ConnectorNotFound,
+
+		/// Unknown connector.
+		UnknownConnector,
+
+		/// Router not found.
+		RouterNotFound,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Set a domain's router,
+		#[pallet::weight(< T as Config >::WeightInfo::set_domain_router())]
+		#[pallet::call_index(0)]
+		pub fn set_domain_router(
+			origin: OriginFor<T>,
+			domain: Domain,
+			router: T::Router,
+		) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			ensure!(domain != Domain::Centrifuge, Error::<T>::DomainNotSupported);
+
+			router.init().map_err(|_| Error::<T>::RouterInitFailed)?;
+
+			<DomainRouters<T>>::insert(domain.clone(), router.clone());
+
+			Self::deposit_event(Event::DomainRouterSet { domain, router });
+
+			Ok(())
+		}
+
+		/// Add a connector for a specific domain.
+		#[pallet::weight(< T as Config >::WeightInfo::add_connector())]
+		#[pallet::call_index(1)]
+		pub fn add_connector(origin: OriginFor<T>, connector: DomainAddress) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				connector.domain() != Domain::Centrifuge,
+				Error::<T>::DomainNotSupported
+			);
+
+			ensure!(
+				!ConnectorsAllowlist::<T>::contains_key(connector.domain(), connector.clone()),
+				Error::<T>::ConnectorAlreadyAdded,
+			);
+
+			ConnectorsAllowlist::<T>::insert(connector.domain(), connector.clone(), ());
+
+			Self::deposit_event(Event::ConnectorAdded { connector });
+
+			Ok(())
+		}
+
+		/// Remove a connector from a specific domain.
+		#[pallet::weight(< T as Config >::WeightInfo::remove_connector())]
+		#[pallet::call_index(2)]
+		pub fn remove_connector(origin: OriginFor<T>, connector: DomainAddress) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin.clone())?;
+
+			ensure!(
+				ConnectorsAllowlist::<T>::contains_key(connector.domain(), connector.clone()),
+				Error::<T>::ConnectorNotFound,
+			);
+
+			ConnectorsAllowlist::<T>::remove(connector.domain(), connector.clone());
+
+			Self::deposit_event(Event::ConnectorRemoved { connector });
+
+			Ok(())
+		}
+
+		/// Process an incoming message.
+		#[pallet::weight(0)]
+		#[pallet::call_index(3)]
+		pub fn process_msg(
+			origin: OriginFor<T>,
+			msg: BoundedVec<u8, T::MaxIncomingMessageSize>,
+		) -> DispatchResult {
+			let domain_address = T::LocalOrigin::ensure_origin(origin)?;
+
+			match domain_address {
```

Comment:
I remember having a similar discussion with Frederik and we decided against abstracting things too much. However, I'm also not a big fan of this coupling with a particular type, but if we change it now we'll also have to change domain address and everything related to it's usage as well.

Reply:
Maybe we can have a `DestinationDomain` type that returns an `EVMChainID` and a `DestinationAddress` type that returns an EVM address?

Reply:
Not sure if I understand you well, do you mean to use 2 different associated types attached to the `connectors-gateway` `Config` trait?

Reply:
Yes, something like that.

Reply:
Ok, I like it.
I was checking a little bit about how both types relate. What about something like this:
```rust
trait Config {
   // Others associated types
   type DomainAddress; // Some basic bounds here that you need
   type Domain: From<DomainAddress>;
}
```

You would need to implement that `From` conversion in `domain_address.rs`, but I think in this way, you can handle both things generically while you still have information about the "relation" among them.

Of course, feel free to leave this for the future if you want.

Reply:
I spent some time looking into this and I'm not sure it's something we want to tackle right now.

If we change the gateway pallet, we change the inbound/outbound queue implementations, thus requiring additional changes in the connectors pallet, which I would prefer to avoid.

You can find a diff of what a portion of the changes would look like [here](https://github.com/centrifuge/centrifuge-chain/compare/connectors-gateway-1-gateway-pallet...gateway-domain-refactor). 

Alternatively, I think we could also add some converters between the types used in the gateway to the types used in the connectors pallet, you can find a smaller changeset [here](https://github.com/centrifuge/centrifuge-chain/compare/connectors-gateway-1-gateway-pallet...gateway-domain-refactor-2)

If we want to go ahead and tackle this now, I would go with the second/shorter option given the lack of knowledge that I have on the connectors pallet itself.

Reply:
I like the idea, but we still use `EVMAddress` because the `EnsureLocal` type uses the type directly. I think `EnsureLocal<ExternalAddress>` or something like that could remove that dependency entirely (?). 

Anyways this issue comes because originally we were not splitting well from types used by the runtime and generic types used by pallets (#1380), and we tend to these issues where we abstract from types while the types are still accessed by the pallets because we need the `cfg-types` dependency.

Yeah, I think we can wait for this change by now until we have a better repo organization. Good ideas anyways! Maybe you can create an issue with those to track this change (?)

Reply:
I think it's fine to postpone this to later. Let's first focus on getting `Connectors` working and then do some cleanups of minor, non-storage related design decisions such as this one.

Reply:
Created an issue for this - https://github.com/centrifuge/centrifuge-chain/issues/1459